### PR TITLE
Mention the specific operator in lint message for vector_logic_linter()

### LINKED ISSUE #2390: vector_logic_linter should reference the specific corresponding operator (`&`->`&&`, `|`->`||`)
This should reference the specific corresponding operator (`&`->`&&`, `|`->`||`)

_Originally posted by @MichaelChirico in https://github.com/r-lib/lintr/pull/2385#discussion_r1414783630_

### PR DESCRIPTION
**Pull request by MichaelChirico**
Closes #2390

### CODE REVIEW

File: R/vector_logic_linter.R [hunk 1]
```diff
@@ -81,15 +81,18 @@ vector_logic_linter <- function() {
...
     xml <- source_expression$xml_parsed_content
     if (is.null(xml)) return(list())
     bad_expr <- xml_find_all(xml, xpath)
 
+    op <- xml_name(bad_expr)
     xml_nodes_to_lints(
       bad_expr,
       source_expression = source_expression,
-      lint_message = "Conditional expressions require scalar logical operators (&& and ||)",
+      lint_message = sprintf("Use `%s` in conditional expressions.", scalar_map[op]),
```
Comment (AshesITR) [inline R/vector_logic_linter.R:95]:
```suggestion
      lint_message = sprintf("Use `%1$s%1$s` in conditional expressions.", op),
```

File: R/vector_logic_linter.R [hunk 2]
```diff
@@ -81,15 +81,18 @@ vector_logic_linter <- function() {
...
   "
 
+  scalar_map <- c(AND = "&&", OR = "||")
+
   Linter(linter_level = "expression", function(source_expression) {
     xml <- source_expression$xml_parsed_content
     if (is.null(xml)) return(list())
     bad_expr <- xml_find_all(xml, xpath)
 
+    op <- xml_name(bad_expr)
```
Comment (AshesITR) [inline R/vector_logic_linter.R:91]:
```suggestion
    op <- xml_text(bad_expr)
```

File: R/vector_logic_linter.R [hunk 3]
```diff
@@ -81,15 +81,18 @@ vector_logic_linter <- function() {
   ]
   "
 
+  scalar_map <- c(AND = "&&", OR = "||")
```
Comment (AshesITR) [inline R/vector_logic_linter.R:84]: We could avoid this map by pasting the op text twice instead.

Comment (MichaelChirico) [reply R/vector_logic_linter.R:84]: that feels over-optimized to me, vector look-up map is very fast in general. If anything I'd do `strrep(xml_text(bad_expr), 2L)` instead, WDYT?

Comment (MichaelChirico) [reply R/vector_logic_linter.R:84]: In any case, the `map` approach is way faster than `sprintf()`, and only slower than `strrep()` in extreme cases where there's a huge # of lints in the same expression:

```r
op_and = "AND"
op_or = "||"
both = sample(c(op_and, op_or), 1e3, TRUE)
map = c(AND = "&&", or = "||")

str_and = "&"
str_or = "|"
str_both = sample(c(str_or, str_and), 1e3, TRUE)

microbenchmark(times = 1e4,
  map[op_and], sprintf("%1$s%1$s", str_and), strrep(str_and, 2L),
  map[op_or], sprintf("%1$s%1$s", str_or), strrep(str_or, 2L),
  map[both], sprintf("%1$s%1$s", str_both), strrep(str_both, 2L))
Unit: nanoseconds
                          expr    min     lq        mean median       uq      max neval cld
                   map[op_and]    421    521    845.0018    581    721.0    98771 10000 a  
  sprintf("%1$s%1$s", str_and)    762    861   1343.1966    941   1150.0   111471 10000 a  
           strrep(str_and, 2L)    710    812   1307.5081    901   1111.0    99381 10000 a  
                    map[op_or]    591    730   2214.5031    831   1061.0 10363474 10000 a  
   sprintf("%1$s%1$s", str_or)    770    870   1330.1354    941   1150.0    98651 10000 a  
            strrep(str_or, 2L)    710    821   1318.9603    901   1121.0    61772 10000 a  
                     map[both]  45481  49541  68498.7497  52581  58176.0 10830374 10000  b 
 sprintf("%1$s%1$s", str_both) 125141 126591 141180.6610 127491 134721.5 10666224 10000   c
          strrep(str_both, 2L)  52001  55821  64786.3580  56250  60006.5 10300914 10000  b
```

I find the look-up approach perfectly readable, but maybe `strrep()` is about the same.

Comment (AshesITR) [reply R/vector_logic_linter.R:84]: Your benchmark is only valid if you copy the full lint message into the map 😉
Anyway, I'd prefer a self-contained approach where all parts of the lint message are assembled in the same place, either outside or inside the linter function.

Comment (MichaelChirico) [reply R/vector_logic_linter.R]: Changed to `strrep()`.

**Review by AshesITR — APPROVED** [review]

**Pull request merged**